### ISSUE TITLE
Tony hawk's Pro Skater 3 (Bad Sound Rendering)

### ISSUE THREAD

**Issue by Catarax** [kode54/vgmstream#33]
Listen to the Redman track from PS2 or Gamecube version of Tony hawk's Pro Skater 3, vgmstream will have a weird sound rendering at 1:55 of the track. Some others tracks seems affected by this at some parts.

PS2:
http://psf2.joshw.info/t/Tony%20Hawk's%20Pro%20Skater%203%20(2001-10-28)(Neversoft)(Activision).7z
GameCube :
http://gcn.joshw.info/t/Tony%20Hawk's%20Pro%20Skater%203%20(2001-11-13)(Neversoft)(Activision).7z

**Comment by kode54** [comment]
I'm more inclined to think this is a bad rip, rather than a problem with the library. I'm guessing it isn't gracefully handling the problem of one channel running out of sample data before the other, due to interleave at the end of the track having an uneven amount of data. I'll have to look closer at the files first, to know for sure.

**Comment by Catarax** [comment]
Well, that's weird to see it on gamecube and ps2 version too. 
Even with a redump verified playstation 2 disc (music.wad), the sound problem is still there, this is not a matter of a bad rip. If you want the files, i can give them to you if necessary.

**Comment by kode54** [comment]
Don't mistake me. I do not mean a bad rip of the game's media to a disc image, and then to its contained data files. I mean a bad rip using whichever stream ripping tools were used to convert the files to the individual songs per file format in the rips available on joshw.info. That would be down to the ripping tools.

**Comment by Catarax** [comment]
Ah okay, well what can I say is that even when you don't use the tools, you get the same sound error when you convert directly the music.wad archive to genh and play it on foobar for example.

This is the only Tony Hawk's episode with this kind of errors, all others episodes from the serie have no issues when i convert the whole .wad mix.

**Comment by Catarax** [comment]
Don't know if it can still be looked at anymore, but the sound of theses files are not properly handled, without ripping tools used (GameCube or Playstation 2 have the same errors while playing).

**Comment by bnnm** [comment]
I tested this with "redman". Around 1.54 (~ 0x5ff200) there are some glitchy-sounding data.
I checked the US and EUR versions of the PS2 game and "redman" is actually slightly different, but both have the glitch. So probably a mastering error by the dev, not a bad rip since the GC ver. has it too.

Now I compared the waveform output of:
- vgmstream default decoding
- vgmstream with PCSX2's decoding
  (https://github.com/PCSX2/pcsx2/blob/master/plugins/spu2-x/src/Mixer.cpp)
- Sony's PC VAGedit (PSYQ devkit)

PCSX2's algorithm is a bit different, but here it just discards the glitch, so you don't hear it as much, but it's still there.
VAGedit *does* have the clipping. However this a program to create/play PS1 ADPCM.

Some possibilities I see:
- PS2 decoding is equal to PS1 decoding, so vgmstream is correct here.
- PS2 decoding is not equal to PS1 decoding, so PCSX2 is correct here.
- VAGedit is also wrong and not hardware accurate.

I think it's worth checking what real hardware does.

Note that vgsmtream decodes with floating point, but the PS1 didn't have an FPU... Also the output is inverted compared to VAGedit's.
